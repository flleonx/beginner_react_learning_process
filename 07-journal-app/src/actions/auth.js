import Swal from 'sweetalert2'
import {
  createUserWithEmailAndPassword,
  signInWithPopup,
  updateProfile,
  signInWithEmailAndPassword,
  signOut
} from 'firebase/auth';
import { auth, googleAuthProvider } from "../firebase/firebaseConfig";
import { types } from "../types/types";
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';

export const startLoginEmailPassword = ( email, password ) =>{
  return async (dispatch) => {
    try {
      dispatch( startLoading() );
      const { user } = await signInWithEmailAndPassword( auth, email, password );
      dispatch( login( user.uid, user.displayName ) );
      dispatch( finishLoading() );
    } catch (error) {
      dispatch( finishLoading() );
      Swal.fire("Error", error.message, "error");
      console.log(error);
    }
  };
};

export const startRegisterWithEmailPasswordName = (name, email, password) => {
  return async ( dispatch ) => {
    try {
      const { user } = await createUserWithEmailAndPassword( auth, email, password );
      await updateProfile( user, { displayName: name });
      dispatch( login(user.uid, user.displayName) );
    } catch (error) {
      Swal.fire("Error", error.message, "error");
      console.log(error);
    }
  };
};

export const startGoogleLogin = () => {
  return async ( dispatch ) => {
    const { user } = await signInWithPopup( auth, googleAuthProvider );
    dispatch( login( user.uid, user.displayName ));
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});

export const startLogout = () => {
  return async ( dispatch ) => {
    await signOut( auth );
    dispatch( logout() );
    dispatch( noteLogout() );
  };
};

export const logout = () => ({
  type: types.logout
});
