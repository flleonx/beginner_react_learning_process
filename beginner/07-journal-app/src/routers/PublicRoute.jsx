import { Redirect, Route } from "react-router-dom";
import PropTypes from "prop-types";

export const PublicRoute = ({ isAuthenticated, component: Component, ...rest }) => {

  return (
    // Each component get some properties that are passed by inference
    <Route { ...rest } component={ (props) => (
      ( isAuthenticated )
        ? ( <Redirect to="/"/> )
        : ( <Component { ...props } /> )
      )}
    />
  )
};

PublicRoute.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  component: PropTypes.func.isRequired,
};
