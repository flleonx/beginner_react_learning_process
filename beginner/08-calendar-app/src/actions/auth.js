import Swal from "sweetalert2";

import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { calendarEventLogout } from "./calendarEvent";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await fetchWithoutToken("auth", { email, password }, "POST");
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user
});

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    const resp = await fetchWithoutToken("auth/new", { name, email, password }, "POST");
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchWithToken("auth/renew");
    const body = await resp.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }));
    } else {
      dispatch(checkingFinish());
    }
  };
};

const checkingFinish = () => ({ type: types.authCheckingFinish });

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(calendarEventLogout());
    dispatch(logout());
  };
};

const logout = () => ({ type: types.authLogout })
