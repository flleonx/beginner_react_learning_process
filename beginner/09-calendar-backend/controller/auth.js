const { response } = require("express");
const bcrypt = require("bcryptjs");
const { generateJWT } = require("../helpers/jwt");

const User = require("../models/User");

const createUser = async (req, res = response) => {

  const { email, password } = req.body;

  try {

    let user = await User.findOne({ email });

    if (user) {
      return res.status(400).json({
        ok: false,
        msg: "There is already a user with that email"
      });
    }

    user = new User(req.body);

    // Password encryption
    const salt = bcrypt.genSaltSync();
    user.password = bcrypt.hashSync(password, salt);

    await user.save();

    // Generate JWT
    const token = await generateJWT(user.id, user.name);

    return res.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name,
      token
    });

  } catch (error) {
    console.log(error);
    return res.status(500).json({
      ok: false,
      msg: "Please talk with the admin",
    });
  }
};

const loginUser = async (req, res = response) => {

  try {

    const { email, password } = req.body;

    const user = await User.findOne({ email });

    // Use a more generic msg
    if (!user) {
      return res.status(400).json({
        ok: false,
        msg: "There is no user with that email"
      });
    };

    // Validate password
    const validPassword = bcrypt.compareSync(password, user.password);

    if (!validPassword) {
      return res.status(400).json({
        ok: false,
        msg: "Wrong password"
      });
    };

    // Generate JWT
    const token = await generateJWT(user.id, user.name);

    return res.json({
      ok: true,
      uid: user.id,
      name: user.name,
      token
    });

  } catch (error) {
    console.log(error);
    return res.status(500).json({
      ok: false,
      msg: "Please talk with the admin",
    });
  }

};

const renewToken = async (req, res = response) => {

  const { uid, name } = req;

  // Generate JWT
  const token = await generateJWT(uid, name);

  return res.status(200).json({
    ok: true,
    uid,
    name,
    token
  });

};

module.exports = {
  createUser,
  loginUser,
  renewToken,
};
