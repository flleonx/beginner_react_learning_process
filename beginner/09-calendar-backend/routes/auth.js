/*
  User routes / Auth
  host + /api/auth
*/
const { Router } = require("express");

const { createUser, loginUser, renewToken } = require("../controller/auth");
const { createUserSchema, loginUserSchema } = require("../models/joi-schemas");
const { loadSchema, fieldsValidator } = require("../middlewares/fields-validator");
const { jwtValidator } = require("../middlewares/jwt-validator");

const auth = Router();

auth.post("/", [
  loadSchema(loginUserSchema),
  fieldsValidator
], loginUser);

auth.post("/new", [
  loadSchema(createUserSchema),
  fieldsValidator
],createUser);

auth.get("/renew", [
  jwtValidator
], renewToken);

module.exports = auth;
