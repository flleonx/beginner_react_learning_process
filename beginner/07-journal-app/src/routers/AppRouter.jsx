import { useEffect, useState } from "react";
import {
  BrowserRouter as Router,
  Switch,
  Redirect,
} from "react-router-dom";
import { onAuthStateChanged } from "firebase/auth";
import { useDispatch } from "react-redux";

import { JournalScreen } from "../components/journal/JournalScreen";
import { AuthRouter } from "./AuthRouter";
import { auth } from "../firebase/firebaseConfig";
import { login } from "../actions/auth";
import { PrivateRoute } from "./PrivateRoute";
import { PublicRoute } from "./PublicRoute";
import { startLoadingNotes } from "../actions/notes";

export const AppRouter = () => {

  const dispatch = useDispatch();

  const [checking, setChecking] = useState(true);
  const [isLoggedIn, setLoggedIn] = useState(false);

  useEffect(() => {
    onAuthStateChanged(auth, async (user) => {
      if (user?.uid) {
        dispatch(login(user.uid, user.displayName));
        setLoggedIn( true );
        dispatch( startLoadingNotes( user.uid ) );
      } else {
        setLoggedIn( false );
      }

      setChecking(false);
    });
  }, [ dispatch ]);

  if ( checking ) {
    return (
      <h1>Wait...</h1>
    )
  }

  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute
            path="/auth"
            component={ AuthRouter  }
            isAuthenticated={ isLoggedIn }
          />
          <PrivateRoute
            exact
            path="/"
            component={ JournalScreen  }
            isAuthenticated={ isLoggedIn }
          />
          <Redirect to="/auth/login" />
        </Switch>
      </div>
    </Router>
  );
};
