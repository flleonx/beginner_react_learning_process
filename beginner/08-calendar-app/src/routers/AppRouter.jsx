import React, { useEffect } from 'react';
import { useDispatch, useSelector } from "react-redux";
import { BrowserRouter as Router, Switch, Redirect } from 'react-router-dom';
import CircularProgress from '@mui/material/CircularProgress';
import Grid from '@mui/material/Grid';
import Container from '@mui/material/Container';

import { CalendarScreen } from '../components/calendar/CalendarScreen';
import { LoginScreen } from '../components/auth/LoginScreen';
import { PublicRoute } from './PublicRoute';
import { PrivateRoute } from './PrivateRoute';

import { startChecking } from "../actions/auth";

export const AppRouter = () => {

  const dispatch = useDispatch();
  const { checking, uid } = useSelector(state => state.auth);

  useEffect(() => {
    setTimeout(() => {
      dispatch(startChecking());
    }, 1000);
  }, [dispatch]);

  if (checking) {
    return (
      <Container style={{alignItems: "center", justifyContent: "center"}}>
      <Grid
        container
        spacing={0}
        direction="column"
        alignItems="center"
        justifyContent="center"
        style={{ minHeight: '100vh' }}
      >
        <Grid item xs={3}>
          <CircularProgress size={300}/>
        </Grid>
      </Grid>
      </Container>
    )
  };

  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute
            exact
            path="/login"
            component={ LoginScreen }
            isAuthenticated={!!uid}
          />
          <PrivateRoute
            exact
            path="/"
            component={ CalendarScreen }
            isAuthenticated={!!uid}
          />
          <Redirect to="/" />
        </Switch>
      </div>
    </Router>
  )
};
