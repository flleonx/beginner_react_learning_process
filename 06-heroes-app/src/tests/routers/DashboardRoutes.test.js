import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";

import { AuthContext } from "../../auth/authContext";
import { DashboardRoutes } from "../../routers/DashboardRoutes";

describe('Tests on <DashboardRoutes />', () => {

  const contextValue = {
    user: {
      logged: true,
      name: "flleonx"
    }
  };

  test('It should show <DashboardRoutes /> properly - Marvel', () => {

    const wrapper = mount(
      <AuthContext.Provider value={ contextValue }>
        <MemoryRouter initialEntries={ ["/"] }>
          <DashboardRoutes />
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find(".text-info").text().trim() ).toBe( "flleonx" );
    expect( wrapper.find("h1").text().trim() ).toBe( "MarvelScreen" );

  });

  test('It should show <DashboardRoutes /> properly - DC', () => {

    const wrapper = mount(
      <AuthContext.Provider value={ contextValue }>
        <MemoryRouter initialEntries={ ["/dc"] }>
          <DashboardRoutes />
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find(".text-info").text().trim() ).toBe( "flleonx" );
    expect( wrapper.find("h1").text().trim() ).toBe( "DCScreen" );

  });

});
