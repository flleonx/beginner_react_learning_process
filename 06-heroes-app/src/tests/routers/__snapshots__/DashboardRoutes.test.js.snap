// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <DashboardRoutes /> It should show <DashboardRoutes /> properly - DC 1`] = `
Array [
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
    <div
      className="navbar-collapse"
    >
      <div
        className="navbar-nav"
      >
        <a
          className="nav-item nav-link "
          href="/marvel"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          aria-current="page"
          className="nav-item nav-link active"
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          className="nav-item nav-link "
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
    </div>
    <div
      className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
    >
      <ul
        className="navbar-nav ml-auto"
      >
        <span
          className="nav-item nav-link text-info"
        >
          flleonx
        </span>
        <button
          className="nav-item nav-link btn"
          onClick={[Function]}
        >
          Logout
        </button>
      </ul>
    </div>
  </nav>,
  <div
    className="container"
  >
    Array [
      <h1>
        DCScreen
      </h1>,
      <hr />,
      <div
        className="row row-cols-1 row-cols-md-3 g-3 animate__animated animate__fadeIn"
      >
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Batman"
                  className="card-img"
                  src="/assets/dc-batman.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Batman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Wayne
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Superman"
                  className="card-img"
                  src="/assets/dc-superman.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Superman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Kal-El
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-superman"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Flash"
                  className="card-img"
                  src="/assets/dc-flash.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Flash
                  </h5>
                  <p
                    className="card-text"
                  >
                    Jay Garrick
                  </p>
                  <p
                    className="text-muted"
                  >
                    Jay Garrick, Barry Allen, Wally West, Bart Allen
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-flash"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Green Lantern"
                  className="card-img"
                  src="/assets/dc-green.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Green Lantern
                  </h5>
                  <p
                    className="card-text"
                  >
                    Alan Scott
                  </p>
                  <p
                    className="text-muted"
                  >
                    Alan Scott, Hal Jordan, Guy Gardner, John Stewart, Kyle Raynor, Jade, Sinestro, Simon Baz
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-green"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Green Arrow"
                  className="card-img"
                  src="/assets/dc-arrow.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Green Arrow
                  </h5>
                  <p
                    className="card-text"
                  >
                    Oliver Queen
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-arrow"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Wonder Woman"
                  className="card-img"
                  src="/assets/dc-wonder.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Wonder Woman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Princess Diana
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-wonder"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Martian Manhunter"
                  className="card-img"
                  src="/assets/dc-martian.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Martian Manhunter
                  </h5>
                  <p
                    className="card-text"
                  >
                    J'onn J'onzz
                  </p>
                  <p
                    className="text-muted"
                  >
                    Martian Manhunter
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-martian"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Robin/Nightwing"
                  className="card-img"
                  src="/assets/dc-robin.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Robin/Nightwing
                  </h5>
                  <p
                    className="card-text"
                  >
                    Dick Grayson
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-robin"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Blue Beetle"
                  className="card-img"
                  src="/assets/dc-blue.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Blue Beetle
                  </h5>
                  <p
                    className="card-text"
                  >
                    Dan Garret
                  </p>
                  <p
                    className="text-muted"
                  >
                    Dan Garret, Ted Kord, Jaime Reyes
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-blue"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Black Canary"
                  className="card-img"
                  src="/assets/dc-black.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Black Canary
                  </h5>
                  <p
                    className="card-text"
                  >
                    Dinah Drake
                  </p>
                  <p
                    className="text-muted"
                  >
                    Dinah Drake, Dinah Lance
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/dc-black"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>,
    ]
  </div>,
]
`;

exports[`Tests on <DashboardRoutes /> It should show <DashboardRoutes /> properly - Marvel 1`] = `
Array [
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
    <div
      className="navbar-collapse"
    >
      <div
        className="navbar-nav"
      >
        <a
          className="nav-item nav-link "
          href="/marvel"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          className="nav-item nav-link "
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          className="nav-item nav-link "
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
    </div>
    <div
      className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
    >
      <ul
        className="navbar-nav ml-auto"
      >
        <span
          className="nav-item nav-link text-info"
        >
          flleonx
        </span>
        <button
          className="nav-item nav-link btn"
          onClick={[Function]}
        >
          Logout
        </button>
      </ul>
    </div>
  </nav>,
  <div
    className="container"
  >
    Array [
      <h1>
        MarvelScreen
      </h1>,
      <hr />,
      <div
        className="row row-cols-1 row-cols-md-3 g-3 animate__animated animate__fadeIn"
      >
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Spider Man"
                  className="card-img"
                  src="/assets/marvel-spider.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Spider Man
                  </h5>
                  <p
                    className="card-text"
                  >
                    Peter Parker
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-spider"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Captain America"
                  className="card-img"
                  src="/assets/marvel-captain.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Captain America
                  </h5>
                  <p
                    className="card-text"
                  >
                    Steve Rogers
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-captain"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Iron Man"
                  className="card-img"
                  src="/assets/marvel-iron.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Iron Man
                  </h5>
                  <p
                    className="card-text"
                  >
                    Tony Stark
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-iron"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Thor"
                  className="card-img"
                  src="/assets/marvel-thor.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Thor
                  </h5>
                  <p
                    className="card-text"
                  >
                    Thor Odinson
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-thor"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Hulk"
                  className="card-img"
                  src="/assets/marvel-hulk.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Hulk
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Banner
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-hulk"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Wolverine"
                  className="card-img"
                  src="/assets/marvel-wolverine.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Wolverine
                  </h5>
                  <p
                    className="card-text"
                  >
                    James Howlett
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-wolverine"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Daredevil"
                  className="card-img"
                  src="/assets/marvel-daredevil.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Daredevil
                  </h5>
                  <p
                    className="card-text"
                  >
                    Matthew Michael Murdock
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-daredevil"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Hawkeye"
                  className="card-img"
                  src="/assets/marvel-hawkeye.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Hawkeye
                  </h5>
                  <p
                    className="card-text"
                  >
                    Clinton Francis Barton
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-hawkeye"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Cyclops"
                  className="card-img"
                  src="/assets/marvel-cyclops.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Cyclops
                  </h5>
                  <p
                    className="card-text"
                  >
                    Scott Summers
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-cyclops"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="col animate__animated animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Silver Surfer"
                  className="card-img"
                  src="/assets/marvel-silver.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-tittle"
                  >
                    Silver Surfer
                  </h5>
                  <p
                    className="card-text"
                  >
                    Norrin Radd
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    />
                  </p>
                  <a
                    href="/hero/marvel-silver"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>,
    ]
  </div>,
]
`;
